packages/flutter/lib/src/rendering/paragraph.dart

```diff
@@ -180,33 +194,48 @@ class RenderParagraph extends RenderBox {
     // us to validate our intrinsic sizes, which will change _textPainter's
     // layout because the intrinsic size calculations are destructive.
     final Size textSize = _textPainter.size;
-    size = constraints.constrain(textSize);
+    final double effectiveLines =
+      // (overflow == TextOverflow.ellipsis && maxLines == null) ? 1.0 :
+      (maxLines?.toDouble() ?? double.INFINITY);
+    final double constrainedHeight =
+      textSize.height.clamp(0.0, _textPainter.preferredLineHeight * effectiveLines);
+    size = constraints.constrain(new Size(textSize.width, constrainedHeight));
 
     final bool didOverflowWidth = size.width < textSize.width;
+    final bool didOverflowHeight = size.height < textSize.height;
     // TODO(abarth): We're only measuring the sizes of the line boxes here. If
     // the glyphs draw outside the line boxes, we might think that there isn't
     // visual overflow when there actually is visual overflow. This can become
     // a problem if we start having horizontal overflow and introduce a clip
     // that affects the actual (but undetected) vertical overflow.
-    _hasVisualOverflow = didOverflowWidth || size.height < textSize.height;
-    if (didOverflowWidth) {
+    _hasVisualOverflow = didOverflowWidth || didOverflowHeight;
+    if (_hasVisualOverflow) {
       switch (_overflow) {
         case TextOverflow.clip:
         case TextOverflow.ellipsis:
           _overflowShader = null;
           break;
         case TextOverflow.fade:
-          TextPainter fadeWidthPainter = new TextPainter(
+          TextPainter fadeSizePainter = new TextPainter(
             text: new TextSpan(style: _textPainter.text.style, text: '\u2026'),
             textScaleFactor: textScaleFactor
           )..layout();
-          final double fadeEnd = size.width;
-          final double fadeStart = fadeEnd - fadeWidthPainter.width;
-          // TODO(abarth): This shader has an LTR bias.
-          _overflowShader = new ui.Gradient.linear(
-            <Point>[new Point(fadeStart, 0.0), new Point(fadeEnd, 0.0)],
-            <Color>[const Color(0xFFFFFFFF), const Color(0x00FFFFFF)]
-          );
+          if (didOverflowWidth) {
+            final double fadeEnd = size.width;
+            final double fadeStart = fadeEnd - fadeSizePainter.width;
+            // TODO(abarth): This shader has an LTR bias.
+            _overflowShader = new ui.Gradient.linear(
+              <Point>[new Point(fadeStart, 0.0), new Point(fadeEnd, 0.0)],
+              <Color>[const Color(0xFFFFFFFF), const Color(0x00FFFFFF)]
+            );
+          } else {
+            final double fadeEnd = size.height;
+            final double fadeStart = fadeEnd - fadeSizePainter.height / 2.0;
+            _overflowShader = new ui.Gradient.linear(
+              <Point>[new Point(0.0, fadeStart), new Point(0.0, fadeEnd)],
+              <Color>[const Color(0xFFFFFFFF), const Color(0x00FFFFFF)]
+            );
```

Comment:
Let's skip the vertical fade.  It's even less clear how it should work than the horizontal fade.